gateway/ln-gateway/src/client.rs

```diff
@@ -69,6 +71,7 @@ impl StandardGatewayClientBuilder {
                 GatewayError::DatabaseError(anyhow::anyhow!("Error opening rocksdb: {e:?}"))
             })?;
             client_builder.with_database(db);
+            client_builder.with_invite_code(invite_code);
```

Comment:
The `StandardGatewayClientBuilder::build` function is called when a gateway is being restarted and is creating clients for already joined federations. Wouldn't this be a problem here since there'd already be a config in the DB?

Reply:
The restart path provides an `old_client` reference, in which case we don't provide an invite code to the client builder

Reply:
It seems to me that `start_gateway` creates an empty `client` map, passes it to `Gateway::new`, which initialized a `Gateway` struct with the empty client map and then calls `Gateway::load_clients` which tried to get an old client from that map, but won't find one and thus passes `None`:

https://github.com/fedimint/fedimint/blob/41e8ce310193408359adda152081ad79a46ee8dc/gateway/ln-gateway/src/lib.rs#L564-L568

I'm very suspicious that tests are running though, so I likely have an error in my thinking somewhere, but that also means the control/data flow is hard to understand. I'll leave the final merge decision to @m1sterc001guy since he has more context.

Reply:
> and then calls Gateway::load_clients which tried to get an old client from that map

`Gateway::load_clients` actually reads from `Gatewayd`'s database, which is why it will actually find a `FederationConfig` there and won't pass `None`. The Lightning reconnect test should be explicitly testing this case, since this is exercised when the connection to the lightning node drops which prompts `Gatewayd` to re-create the `Gateway` struct and re-build clients. So if that test is passing I think we're good. 

Maybe we should add a comment to `load_clients` or rename the `clients` map? That might make this a little easier to understand.

Reply:
> It seems to me that `start_gateway` creates an empty `client` map, passes it to `Gateway::new`, which initialized a `Gateway` struct with the empty client map and then calls `Gateway::load_clients` which tried to get an old client from that map, but won't find one and thus passes `None`:
> 
> https://github.com/fedimint/fedimint/blob/41e8ce310193408359adda152081ad79a46ee8dc/gateway/ln-gateway/src/lib.rs#L564-L568
> 
> I'm very suspicious that tests are running though, so I likely have an error in my thinking somewhere, but that also means the control/data flow is hard to understand. I'll leave the final merge decision to @m1sterc001guy since he has more context.

utACK.

> The Lightning reconnect test should be explicitly testing this case, since this is exercised when the connection to the lightning node drops which prompts Gatewayd to re-create the Gateway struct and re-build clients. So if that test is passing I think we're good.

To recreate the fail scenario @elsirion described above, we need to [rerun Gatewayd with the same db path](https://github.com/fedimint/fedimint/blob/e29d7aac010cefef2ea817bc0467def1556dbca5/gateway/ln-gateway/src/lib.rs#L192C1-L208C11). Lightning tests don't presently cover this scenario

Reply:
> To recreate the fail scenario @elsirion described above, we need to [rerun Gatewayd with the same db path](https://github.com/fedimint/fedimint/blob/e29d7aac010cefef2ea817bc0467def1556dbca5/gateway/ln-gateway/src/lib.rs#L192C1-L208C11). Lightning reconnect tests don't presently cover this scenario 

Not necessary. The bug and untested scenario occurs because we can rebuild gateway fed clients with the same db, but `old_client` is not a sufficient flag to tell that we are in a rebuild scenario, where db possibly has a client config already.

Fix is to explicitly check if config is in client database, using `get_config_from_db`

Reply:
> The bug and untested scenario occurs because we can rebuild gateway fed clients with the same db, but old_client is not a sufficient flag to tell that we are in a rebuild scenario, where db possibly has a client config already.

Is this because we construct `old_client` based on whether or not there's a `FederationConfig` in the database, when we should actually be looking for `ClientConfig`? I'm a bit confused what the gap is here.

Reply:
> I haven't fully tested this, so marking PR as draft

PR #3241 adds test coverage for gateway reboot scenario, which helps us maintain some sanity around gateway client-building on gatewayd boot.